### PR TITLE
Split off ArgumentShuffleExpr from TupleShuffleExpr

### PR DESCRIPTION
**Pull request by slavapestov**
Right now we use TupleShuffleExpr for two completely different things:
  
- Tuple conversions, where elements can be re-ordered and labels can be
  introduced/eliminated
- Complex argument lists, involving default arguments or varargs
  
The first case does not allow default arguments or varargs, and the
second case does not allow re-ordering or introduction/elimination
of labels. Furthermore, the first case has a representation limitation
that prevents us from expressing tuple conversions that change the
type of tuple elements.
  
For all these reasons, it is better if we use two separate Expr kinds
for these purposes.
  
In sequent commits I plan on redesigning TupleShuffleExpr to correctly
represent all tuple conversions without any unnecessary baggage.
  
Longer term, we actually want to change the representation of CallExpr
to directly store an argument list; then instead of a single child
expression that must be a ParenExpr, TupleExpr or ArgumentShuffleExpr,
all CallExprs will have a uniform representation and ArgumentShuffleExpr
will go away altogether. This should reduce memory usage and radically
simplify parts of SILGen.

(Progress on rdar://12340004, https://bugs.swift.org/browse/SR-2672.)

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@xedin This may interest you as well.

File: lib/AST/ASTDumper.cpp [hunk 1]
```diff
@@ -2118,15 +2118,26 @@ class PrintExpr : public ExprVisitor<PrintExpr> {
...
+    for (unsigned i = 0, e = E->getElementMapping().size(); i != e; ++i) {
+      if (i) OS << ", ";
+      OS << E->getElementMapping()[i];
+    }
+    OS << "]\n";
+    printRec(E->getSubExpr());
+    PrintWithColorRAII(OS, ParenthesisColor) << ')';
+  }
+  void visitArgumentShuffleExpr(ArgumentShuffleExpr *E) {
     printCommon(E, "tuple_shuffle_expr");
```
Comment (Azoy) [inline lib/AST/ASTDumper.cpp]: Did you mean to also update this to `argument_shuffle_expr`?

Comment (slavapestov) [reply lib/AST/ASTDumper.cpp]: Good catch. I grepped for TupleShuffleExpr but forgot to check other spellings. I'll fix this in a follow-up PR.

**Pull request merged**